### PR TITLE
add private API for generating all CompositeImplicit decomps from the dispatcher

### PR DESCRIPTION
**Pull request by bdhirsh**
What do people think of this API / location?

Part of the reason it's a bit ugly is because there are 3 different places that decomp can live - the C++ dispatcher, the Python dispatcher, and the python global decomp table.

This should be useful to QAT (cc @andrewor14) - the idea is that they want to be able to use pre_dispatch tracing to trace out an ATen graph above the dispatcher, but they want an easy way to be able to say "specify a decomp table that includes every CompositeImplicitAutograd op that would have decomposed if we had traced through the dispatcher".

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105240
* __->__ #106330
* #106329
* #105865

### CODE REVIEW

File: torch/_ops.py [hunk 1]
```diff
@@ -853,6 +853,20 @@ def has_composite_implicit_kernel(op):
         return False
     return True
 
+# Private API that returns a dict containing all ATen operators
+# that have CompositeImplicitAutograd decompositions registered in the dispatcher.
+def _get_composite_implicit_autograd_decomps() -> Dict[OpOverload, Callable]:
+    import re
+    # Get all C++ CompositeImplicitAutograd decomps.
+    op_names = [re.split(r'::|\.', x) for x in torch._C._dispatch_get_registrations_for_dispatch_key("CompositeImplicitAutograd")]
```
Comment (ezyang) [inline torch/_ops.py:861]: sigh lol

Comment (ezyang) [reply torch/_ops.py:861]: You have two occurrences of this, can we at least have a helper function lol

**Comment by ezyang** [comment]
I'm still not sure why this is a thing they want. @andrewor14 can you explain more?

**Comment by andrewor14** [comment]
Yeah so basically exporting with `pre_dispatch=True` completely changes the opset quantization has to deal with, since the implicit autograd decomps are not run. If we want to mimic the old opset produced when `pre_dispatch=False`, we will need a way to manually specify the same decomps. This is important for making the quantization and delegation opsets similar in the end-to-end flow, so the backend developers don't have to deal with two completely different ones. (More details in the doc)

File: torch/_ops.py [hunk 2]
```diff
@@ -853,6 +853,20 @@ def has_composite_implicit_kernel(op):
...
+    composite_implicit_ops = [getattr(getattr(getattr(torch.ops, x[0]), x[1]), "default" if len(x) < 3 else x[2]) for x in op_names if x[0] == "aten"]
+
+    # The API above only finds ops with a C++ CompositeImplicitAutograd registration.
+    # We also need to get at any ops that were given CompositeImplicitAutograd registrations from python.
+    all_op_names = [re.split(r'::|\.', x) for x in torch._C._dispatch_get_all_op_names()]
+    all_ops = [getattr(getattr(getattr(torch.ops, x[0]), x[1]), "default" if len(x) < 3 else x[2]) for x in all_op_names if x[0] == "aten"]
+    python_composite_implicit_ops = [x for x in all_ops if torch._C.DispatchKey.CompositeImplicitAutograd in x.py_kernels]
```
Comment (kimishpatel) [inline torch/_ops.py:868]: But this gets only those decomps that have py_kernels, right? Those with c++ implementations you cannot get, right? Is that ok and does it provide enough coverage?

Comment (bdhirsh) [reply torch/_ops.py:868]: Actually, line 862 above gets all the C++ composite implicit kernels. This line gets the py_kernels that are also composite implicit, and I cat them together on line 869

File: torch/_ops.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_ops.py:861]: Agreed haha

**Comment by ezyang** [comment]
@andrewor14 If you all prefer the post composite implicit autograd set, then it seems to me that you should have the reverse API: by default these decomps apply even pre_dispatch, and you can block some of them if you want them. This prevents you from having to generate the list of all the decomps.

**Comment by kimishpatel** [comment]
@andrewor14 I think we will discuss this perhaps in today's composability meeting, but it is also looking like that the IR quant works on might be different from what delegate works on.

File: torch/_ops.py (hunk 2 above)
Comment (kimishpatel) [reply torch/_ops.py:868]: Oh thats right. I missed that. But my question still is, how do you get some op X registered with `CompositeImplicitAutograd` key, has decomposition that consist of Y and Z ops. Now that I look at get_decompositions implementation, I see callable is returned. So I presume, make_fx will actually run the callable when it encounters the op in decomp list

**Comment by andrewor14** [comment]
> @andrewor14 If you all prefer the post composite implicit autograd set, then it seems to me that you should have the reverse API: by default these decomps apply even pre_dispatch, and you can block some of them if you want them. This prevents you from having to generate the list of all the decomps.

Yes, that would work for us. If I understand correctly from QAT's perspective this just means we don't need to explicitly pass in this extra implicit autograd decomp table. I think it all depends on what the other expected use cases of export are. If we think the implicit autograd decomps are a common thing that a lot of people will want then this make senses.

On the other hand passing in a decomp table as in this PR is also not that bad if we end up wrapping the export API in a helper thin wrapper, since all the details are hidden from the user anyway. I'd like to discuss this in the composability meeting today.

**Comment by andrewor14** [comment]
> @andrewor14 I think we will discuss this perhaps in today's composability meeting, but it is also looking like that the IR quant works on might be different from what delegate works on.

Yeah it seems the two IRs will necessarily diverge, so I guess the question is the extent. Decomposed vs not decomposed at all will be wildly different, but decomposed with implicit autograd vs core aten will be a lot more similar.

**Comment by bdhirsh** [comment]
@ezyang made a pretty good point that so far, it seems like all the major use cases of pre_dispatch that we care about today do prefer to run the CompositeImplicitAutograd decompositions, which is an argument for making pre_dispatch run them automatically.

The counterargument is mostly that if we start running these decomps by default, we'd now need pre_dispatch to take in a **new** argument specifying decomps that you do not want to run (e.g. `aten.linear`) - if pre_dispatch gets folded into an export API at some point, we'll need to figure out a reasonable way to expose that knob too. But that doesn't seem too unreasonable.

So for now I probably won't land this PR. But @andrewor14 let me know if you want these decomps running with pre_dispatch in main soon, and if everyone's in agreement then we can work on changing the pre_dispatch default to just always run these decompositions.

**Comment by kimishpatel** [comment]
> argument specifying decomps that you do not want to run

@bdhirsh I am not too convinced that this knob is something that should be exposed for the API we are discussing. My concern is that export being quant specific. I think it is good keep pre_dispatch export that has consistent opset that all users of pre_dispatch export can target. If that means dont decompose linear, thats fine, but special casing that for quant does not feel right to me.

**Comment by bdhirsh** [comment]
@kimishpatel sounds good, happy to hold off on any changes until we get in agreement.

> My concern is that export being quant specific

To be fair, I don't think that having pre_dispatch run these decomps by default would be quantization specific: mostly since we know of at least one other pre_dispatch customer that will also want these decomps to run: inductor.

**Comment by kimishpatel** [comment]
> To be fair, I don't think that having pre_dispatch run these decomps by default would be quantization specific: mostly since we know of at least one other pre_dispatch customer that will also want these decomps to run: inductor.

Right. I was referring to the part of "we'd now need pre_dispatch to take in a new argument specifying decomps that you do not want to run (e.g. aten.linear) ". If not decomposing linear is quant specific, do we want such pre_dispatch export path. I am a bit hesitant, but there might be counter arguments to it.

**Comment by andrewor14** [comment]
> So for now I probably won't land this PR. But @andrewor14 let me know if you want these decomps running with pre_dispatch in main soon, and if everyone's in agreement then we can work on changing the pre_dispatch default to just always run these decompositions.

Sounds good to me. I just discussed with Jerry about the IR we want for quantization and I think we do still want these decomps. Is it easy to do these decomps automatically when `pre_dispatch=True`? It would be great for QAT to have this soon.

**Comment by bdhirsh** [comment]
@andrewor14 let me know if this sounds reasonable to you:

(1) If we want to unblock QAT so they can use these decomps asap, we can always land this PR but move the API I added here to be a private util in the quantization namespace (that we can kill later)

(2) We also... could totally change `pre_dispatch` to always run CompositeImplicit decomps immediately. The only thing I'm worried about is that as soon as someone comes to us asking for a way to not decompose e.g. `aten.linear`, we'll hit the issue of "how can we expose that capability in the export API?". Which makes me think that we might want to align a bit more on how this factors into the export API before flipping the default in pre_dispatch.

**Comment by kimishpatel** [comment]
> as someone comes to us asking for a way to not decompose e.g. aten.linear, we'll hit the issue of "how can we expose that capability in the export API?".

Brian, isnt that true for all export APIs? That is some decomps are not what people may expect?

**Comment by jerryzh168** [comment]
@bdhirsh just to confirm this PR stack is blocked on quantization's plan of which opset we want to use right

**Comment by bdhirsh** [comment]
@jerryzh168 yep, I think that's the case

**Pull request closed without merging**